kedro/contrib/io/pyspark/spark_hive_data_set.py

```diff
@@ -0,0 +1,306 @@
+        >>> data_set.save(spark_df)
+        >>> reloaded = data_set.load()
+        >>>
+        >>> reloaded.take(4)
+    """
+
+    def _describe(self) -> Dict[str, Any]:
+        return dict(
+            database=self._database,
+            table=self._table,
+            write_mode=self._write_mode,
+            table_pk=self._table_pk,
+        )
+
+    def __init__(
+        self, database: str, table: str, write_mode: str, table_pk: List[str] = None
+    ) -> None:
+        """Creates a new instance of ``SparkHiveDataSet``.
+
+        Args:
+            database: The name of the hive database.
+            table: The name of the table within the database.
+            write_mode: ``insert``, ``upsert`` or ``overwrite`` are supported.
+            table_pk: If performing an upsert, this identifies the primary key columns used to
+                resolve preexisting data. Is required for ``write_mode="upsert"``
+
+        Raises:
+            DataSetError: Invalid configuration supplied
+        """
+        self._database = database
+        self._table = table
+        self._stage_table = "_temp_" + table
+        self._valid_write_modes = ["insert", "upsert", "overwrite"]
+        if write_mode not in self._valid_write_modes:
+            raise DataSetError(
+                "Invalid write_mode provided: {invalid}. Write_mode must be one of {valid}".format(
+                    invalid=write_mode, valid=self._valid_write_modes
+                )
+            )
+        self._write_mode = write_mode
+        if self._write_mode == "upsert" and not table_pk:
+            raise DataSetError("table_pk must be set to utilise upsert read mode")
+        self._table_pk = table_pk
+
+        self._table_columns = self._load().columns if self._exists() else None
+
+        if (
+            self._table_pk
+            and self._exists()
+            and not all(
+                (col_name in self._table_columns for col_name in self._table_pk)
+            )
+        ):
+            raise DataSetError(
+                "columns [{colnames}] selected as PK not found in table {database}.{table}".format(
+                    colnames=", ".join(self._table_pk),
+                    database=self._database,
+                    table=self._table,
+                )
+            )
+
+    @staticmethod
+    def _get_spark():
+        return SparkSession.builder.getOrCreate()
+
+    def _create_empty_hive_table(self, data):
+        data.createOrReplaceTempView("tmp")
+        self._get_spark().sql(
+            "create table {database}.{table} select * from tmp limit 1".format(
+                table=self._table, database=self._database
+            )
+        )
+        self._get_spark().sql(
+            "truncate table {database}.{table}".format(
+                database=self._database, table=self._table
+            )
+        )
+
+    def _load(self) -> DataFrame:
+        if not self._exists():
+            raise DataSetError(
```

Comment:
Is this check necessary? Will `self._get_spark().sql()` below fail if the table doesn't exist?

Reply:
yea this just gives a custom error rather than a sql analysis error on `self._get_spark().sql()` - both would achieve the same thing but this is perhaps a little more specific? could remove for simplicity, don't mind either way.

Reply:
I'm fine with raising a custom error as long as it provides a better error message